File: lib/postrunner/TrackView.rb
#!/usr/bin/env ruby -w
# encoding: UTF-8
#
# = TrackView.rb -- PostRunner - Manage the data from your Garmin sport devices.
#
# Copyright (c) 2014 by Chris Schlaeger <cs@taskjuggler.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#

require 'fit4ruby'

require 'postrunner/ViewFrame'

module PostRunner

  # A TrackView object uses OpenLayers to draw a map with the waypoints of the
  # activity. Lap end are marked with enumerated lap markers.
  class TrackView

    # Create a new TrackView object for a given Activity.
    # @param activity [Activity] The activity to render
    def initialize(activity)
      @activity = activity
      @session = @activity.fit_activity.sessions[0]
      @has_geo_data = @session.has_geo_data?
    end

    # Render the map widget with the track and marker overlay as HTML code.
    # @param doc [HTMLBuilder] HTML document to add to.
    def to_html(doc)
      return unless @has_geo_data

      doc.head {
        doc.unique(:trackview_style) {
          doc.style(style)
          doc.link({ 'rel' => 'stylesheet',
                     'href' => 'openlayers/ol.css',
                     'type' => 'text/css' })
          doc.script({ 'src' => 'openlayers/ol.js' })
          doc.script({ 'src' => 'postrunner/trackview.js' })
        }
        doc.script(java_script)
      }

      ViewFrame.new('Map', 600) {
        doc.div({ 'id' => 'map', 'class' => 'trackmap' })
      }.to_html(doc)
    end

    private

    def style
      <<EOT
.olControlAttribution {
    bottom: 5px;
}

.trackmap {
  width: 565px;
  height: 400px;
  border: 2px solid #545454;
}
EOT
    end

    def java_script
      center_long = @session.swc_long +
        (@session.nec_long - @session.swc_long) / 2.0
      center_lat = @session.swc_lat +
        (@session.nec_lat - @session.swc_lat) / 2.0

      <<"EOT"
#{track_points}
#{lap_markers}
function init() {
  pr_trackview_init(#{center_long}, #{center_lat});
}
EOT
    end

    # Generate a javascript variable with an Array with the coordinates of the
    # track points. Each coordinate is an Array with a longitude and latitude
    # in EPSG:4326. Generate a javascript variable with an Array of track
    # points.
    def track_points
      points = []
      @activity.fit_activity.sessions.map do |session|
        session.records.map do |record|
          long = record.position_long
          lat = record.position_lat
          next unless long && lat

          points << "[ #{long}, #{lat} ]"
        end
      end

      "var pr_track_points = [\n" +
      points.join(', ') +
      "\n];\n"
    end

    # Generate a javascript variable with an Array with the coordinates of the
    # start point and the lap end points. Each coordinate is an Array with a
    # longitude and latitude in EPSG:4326.
    def lap_markers
      "var pr_lap_markers = [\n" +
        "[ #{@session.start_position_long}," +
        " #{@session.start_position_lat} ], " +
        @activity.fit_activity.laps.map do |lap|
          "[ #{lap.end_position_long}, #{lap.end_position_lat} ]"
        end.join(', ') +
        "\n];\n"
    end

  end

end

File: misc/postrunner/trackview.js
/*
 * TrackView.js -- PostRunner - Manage the data from your Garmin sport devices.
 *
 * Copyright (c) 2014, 2015, 2016 by Chris Schlaeger <cs@taskjuggler.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 */

var pr_trackview_init = function(center_long, center_lat) {
  lm_w = 40;
  lm_h = 100;

  pr_transformer = function(coords, i, arr) {
    arr[i] = ol.proj.transform(coords, 'EPSG:4326', 'EPSG:900913');
  };

  pr_track_points.forEach(pr_transformer);
  pr_lap_markers.forEach(pr_transformer);

  var map = new ol.Map({
    view: new ol.View({
      center: ol.proj.transform([ center_long, center_lat ],
                                'EPSG:4326', 'EPSG:900913'),
      zoom: 14,
    }),
    layers: [
      new ol.layer.Tile({
        source: new ol.source.MapQuest({layer: 'osm'})
      }),
      new ol.layer.Vector({
        source: source_vector = new ol.source.Vector({
          features: [
            new ol.Feature({
              geometry: new ol.geom.LineString(pr_track_points)
            })
          ]
        }),
        style: [
          new ol.style.Style({
            stroke: new ol.style.Stroke({
              color: 'red',
              width: 5
            }),
            fill: new ol.style.Fill({
              color: 'white'
            })
          })
        ]
      }),
      new ol.layer.Vector({
        source: lap_marker_source = new ol.source.Vector(),
        style: function(feature, resolution) {
          return [
            new ol.style.Style({
              stroke: new ol.style.Stroke({
                color: 'black',
                width: 2
              }),
              fill: new ol.style.Fill({
                color: feature.get('color')
              }),
              text: new ol.style.Text({
                font: '' + (lm_w / resolution) + 'px helvetica,sans-serif',
                text: resolution < (lm_w / 8.0) ? feature.get('name') : '',
                fill: new ol.style.Fill({
                  color: 'black'
                })
              })
            })
          ];
        }
      })
    ],
    target: "map"
  });
  for (var i in pr_lap_markers) {
    x = pr_lap_markers[i][0];
    y = pr_lap_markers[i][1];
    lap_marker_source.addFeature(
      new ol.Feature({
        geometry: new ol.geom.Polygon([[[ x - lm_w, y + lm_h ],
                                        [ x, y ],
                                        [ x + lm_w, y + lm_h ],
                                        [ x - lm_w, y + lm_h ]]])
      })
    );
    lap_marker_source.addFeature(
      new ol.Feature({
        geometry: new ol.geom.Circle([ x, y + lm_h ], lm_w),
        name: (i == 0 ? 'S' : i == pr_lap_markers.length - 1 ? 'F' : i),
        color: (i == 0 ? 'green' :
		         i == pr_lap_markers.length - 1 ? 'red' : 'yellow'),
      })
    );
  };
};

